clang/lib/CodeGen/ItaniumCXXABI.cpp

```diff
@@ -3565,8 +3565,28 @@ void ItaniumRTTIBuilder::BuildVTablePointer(const Type *Ty) {
   // Check if the alias exists. If it doesn't, then get or create the global.
   if (CGM.getItaniumVTableContext().isRelativeLayout())
     VTable = CGM.getModule().getNamedAlias(VTableName);
-  if (!VTable)
-    VTable = CGM.getModule().getOrInsertGlobal(VTableName, CGM.Int8PtrTy);
+
+  auto VTableTy =
+      CGM.getLangOpts().SYCLIsDevice &&
+              CGM.getLangOpts().SYCLAllowVirtualFunctions &&
+              (VTableName == ClassTypeInfo || VTableName == SIClassTypeInfo)
+          ? CGM.Int8Ty->getPointerTo(
+                static_cast<unsigned>(LangAS::opencl_global))
+          : CGM.Int8PtrTy;
+  if (!VTable) {
+    if (VTableTy == CGM.Int8PtrTy) {
```

Comment:
VTableTy is `Int8PtrTy` for both SYCL and non-SYCL case above right? So what is the `else` for?

Reply:
I believe `Int8PtrTy` and `Int8PtrTy(LangAS::opencl_global)` are different, am I wrong?

Reply:
I am not sure but in any case, I think it is better to guard this using `SYCLIsDevice`. 

I am not familiar with this code so I may be wrong but do we need the call back function? Won't global variable be constructed with modified VTableTy without the callback back? Or do default constructors not consider address space?

Reply:
I think all new changes need to be guarded with `SYCLIsDevice` if as well as `SYCLAllowVirtualFunctions`.

Reply:
Smth like 
```
if (!VTable) {
  Type *VTableTy = CGM.Int8PtrTy;
  if (CGM.getLangOpts().SYCLIsDevice &&
      CGM.getLangOpts().SYCLAllowVirtualFunctions &&
      (VTableName == ClassTypeInfo || VTableName == SIClassTypeInfo)) {
    VTableTy =
        CGM.Int8Ty->getPointerTo(static_cast<unsigned>(LangAS::opencl_global))
            VTable =
            CGM.getModule().getOrInsertGlobal(VTableName, VTableTy, [&] {
              return new llvm::GlobalVariable(
                  CGM.getModule(), VTableTy, false,
                  llvm::GlobalVariable::ExternalLinkage, nullptr, VTableName,
                  nullptr, llvm::GlobalValue::ThreadLocalMode::NotThreadLocal,
                  llvm::Optional<unsigned>(
                      static_cast<unsigned>(LangAS::opencl_global)));
            });
  } else {
    VTable = CGM.getModule().getOrInsertGlobal(VTableName, VTableTy);
  }
}
```